knowledge-base/grid-show-confirm-cancel-dialog.md

```diff
@@ -0,0 +1,236 @@
+    }
+
+    // in a real case, keep the models in dedicated locations, this is just an easy to copy and see example
+    public class SampleData
+    {
+        public int ID { get; set; }
+        public string Name { get; set; }
+    }
+}
+````
+
+### Dialog Component
+
+Using the [Dialog component]({%slug dialog-overview%}) will let you have fully customized Cancel Confirmation Dialog. To handle the scenario:
+
+* Declare a Dialog instance and add the desired content and buttons there. Normally, you would need at least one button - for confirmation of the cancel operation.
+* Handle the [`OnCancel`]({%slug grid-events%}#cud-events) event of the Grid to show the custom Dialog.
+* Wait for the user choice to prevent or proceed with the cancel operation.
+* Hide the Dialog.
+
+````CSTHML
+@*Only OnCancel event is handled for brevity*@
+
+@using System.ComponentModel.DataAnnotations @* for the validation attributes *@
+
+<TelerikGrid Data=@GridData
+             EditMode="@GridEditMode.Inline"
+             Pageable="true"
+             Sortable="true"
+             Height="500px"
+             OnCancel="@CancelHandler"
+             ShowColumnMenu="true">
+    <GridToolBar>
+        <GridCommandButton Command="Add" Icon="add">Add Employee</GridCommandButton>
+    </GridToolBar>
+    <GridColumns>
+        <GridColumn Field=@nameof(SampleData.ID) Title="ID" Editable="false" />
+        <GridColumn Field=@nameof(SampleData.Name) Title="Name" />
+        <GridCommandColumn>
+            <GridCommandButton Command="Save" Icon="save" ShowInEdit="true">Update</GridCommandButton>
+            <GridCommandButton Command="Edit" Icon="edit">Edit</GridCommandButton>
+            <GridCommandButton Command="Cancel" Icon="cancel" ShowInEdit="true">Cancel</GridCommandButton>
+        </GridCommandColumn>
+    </GridColumns>
+</TelerikGrid>
+
+<TelerikDialog @bind-Visible="@DialogVisible"
+               Title="@Title">
+    <DialogContent>
+        Are you sure you want to cancel the editing?
+    </DialogContent>
+    <DialogButtons>
+        <TelerikButton OnClick="@(() => CancelConfirmation("continueEdit"))">No, continue edit</TelerikButton>
+        <TelerikButton OnClick="@(() => CancelConfirmation("cancelEdit"))">Yes, cancel edit</TelerikButton>
+    </DialogButtons>
+</TelerikDialog>
+
+@code {
+    private List<SampleData> GridData { get; set; }
+
+    private bool DialogVisible { get; set; }
+
+    private string Title { get; set; } = "Please confirm";
+
+    private bool? ContinueEdit { get; set; }
+
+    private async Task CancelHandler(GridCommandEventArgs args)
+    {
+        SampleData item = (SampleData)args.Item;
+
+        //clean the value, so you can always get the updated one
+        ContinueEdit = null;
+
+        //show dialog
+        DialogVisible = true;
+
+        //wait for the user choice from the dialog handled in CancelConfirmation()
+        while (!ContinueEdit.HasValue)
+        {
+            await Task.Delay(10);
```

Comment:
You can use a larger timeout period for more efficient app performance. Users can't distinguish 10ms from 50ms anyway. We use 20ms for the built-in predefined dialogs. See `DialogBuilder.razor.cs`.